### PR TITLE
Add compression codec configurations for data in exchanges and spills

### PR DESCRIPTION
**Pull request by hackeryang**
## Description
Our community knows that ZSTD compression is better about speed and compression ratio in many scenarios, for example:  https://github.com/trinodb/trino/issues/10058  
Many blogs and test reports on the internet also mentioned that ZSTD is better than ZLIB in most cases:  https://databento.com/blog/zstd-vs-zlib

We found that if page serialization and deserialization also use ZSTD, the **compression and decompression performance** is also better, and `Fault Tolerant Execution` also uses compressed pages:  https://github.com/trinodb/trino/pull/14957

The benchmark result is shown as below, It seems that if compression is enabled, Zstd is about 10X faster than Lz4.  

Because ZSTD may not be a best choice for all scenarios, and we want to take into account the demands of different users, we **added a session property and configuration to switch between several compresion codecs, and still use Lz4 by default** when compression is enabled.

### Lz4 Compression
```
Benchmark                        (compressed)  (encrypted)  (randomSeed)   Mode  Cnt     Score     Error  Units
BenchmarkPagesSerde.deserialize          true         true          1000  thrpt   10   104.194 ±   4.508  ops/s
BenchmarkPagesSerde.deserialize          true        false          1000  thrpt   10  4770.274 ± 106.577  ops/s
BenchmarkPagesSerde.deserialize         false         true          1000  thrpt   10    83.668 ±   3.172  ops/s
BenchmarkPagesSerde.deserialize         false        false          1000  thrpt   10  8175.870 ± 423.283  ops/s
BenchmarkPagesSerde.serialize            true         true          1000  thrpt   10    90.786 ±  18.355  ops/s
BenchmarkPagesSerde.serialize            true        false          1000  thrpt   10   952.958 ±  31.424  ops/s
BenchmarkPagesSerde.serialize           false         true          1000  thrpt   10    85.247 ±   1.277  ops/s
BenchmarkPagesSerde.serialize           false        false          1000  thrpt   10  8060.584 ± 119.043  ops/s
```
### Zstd Compression
```
Benchmark                        (compressed)  (encrypted)  (randomSeed)   Mode  Cnt      Score      Error  Units
BenchmarkPagesSerde.deserialize          true         true          1000  thrpt   10    102.974 ±    1.746  ops/s
BenchmarkPagesSerde.deserialize          true        false          1000  thrpt   10    373.737 ±   55.635  ops/s
BenchmarkPagesSerde.deserialize         false         true          1000  thrpt   10     85.315 ±    0.897  ops/s
BenchmarkPagesSerde.deserialize         false        false          1000  thrpt   10  10393.729 ± 1296.839  ops/s
BenchmarkPagesSerde.serialize            true         true          1000  thrpt   10     44.524 ±    0.103  ops/s
BenchmarkPagesSerde.serialize            true        false          1000  thrpt   10     65.594 ±    0.373  ops/s
BenchmarkPagesSerde.serialize           false         true          1000  thrpt   10     83.790 ±    2.491  ops/s
BenchmarkPagesSerde.serialize           false        false          1000  thrpt   10   7675.937 ±  347.427  ops/s
```
Benchmarks of exchanges between nodes in a real cluster on AWS, thanks to Mr Raunaq: https://github.com/trinodb/trino/pull/20274#discussion_r1442633339  
Our discussion about when to use ZSTD: https://trinodb.slack.com/archives/CP1MUNEUX/p1704978337669429  
Snappy VS ZSTD:  https://www.percona.com/blog/compression-methods-in-mongodb-snappy-vs-zstd/  

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# General
* The configuration property `exchange.compression-enabled` is now replaced by `exchange.compression-codec` which allows specifying `NONE`, `LZ4` or `ZSTD` as the compression codec for data exchanged between workers. The corresponding system session property `exchange_compression` is replaced by `exchange_compression_codec`. ({issue}`20274`)
* The configuration property `spill-compression-enabled` is now replaced by `spill-compression-codec` which allows specifying `NONE`, `LZ4` or `ZSTD` as the compression codec for data spilled to disk. The already deprecated configuration property `experimental.spill-compression-enabled` is now defunct and must be removed from server configurations. ({issue}`20274`)
```

### CODE REVIEW

**Comment by hackeryang** [comment]
Hello @findepi @wendigo @losipiuk , can you please take a review when you have time, thank you~

**Comment by wendigo** [comment]
I'm concerned that zstd is using jni which could affect gc @hackeryang

**Comment by wendigo** [comment]
I don't think that we should switch by default. We could make it configurable

**Comment by findepi** [comment]
cc @raunaqmorarka

**Comment by raunaqmorarka** [comment]
> I'm concerned that zstd is using jni which could affect gc @hackeryang

This is using the airlift implementation which is pure java code rather than jni unless I'm mistaken about that

**Comment by raunaqmorarka** [comment]
> Our community knows that ZSTD compression is better about speed and compression ratio in most scenarios, for example: https://github.com/trinodb/trino/issues/10058

That's a comparison between ZSTD and GZIP, both these prioritise achieving higher compression over CPU efficiency. ZSTD is better between these two, that does not automatically make it the better choice for every scenario compared to lighter compression schemes like SNAPPY and LZ4.
ZSTD makes sense for data files or exchange data stored on S3. But it's not clear that it would be the best choice for data exchanged over the network. With a high speed and bandwidth network we're better off using less CPU heavy compression. The existing LZ4 exchange compression wasn't turned on for pipelined execution because its impact is workload and setup dependent, it makes some queries faster and some slower.
This needs benchmarking in an actual Trino cluster rather than JMH or generic compression/decompression speed tests to assess properly.

**Comment by wendigo** [comment]
@raunaqmorarka I've assumed that zstd codec is using zstd-jni underneath, my mistake

**Comment by hackeryang** [comment]
> > Our community knows that ZSTD compression is better about speed and compression ratio in most scenarios, for example: #10058
> 
> That's a comparison between ZSTD and GZIP, both these prioritise achieving higher compression over CPU efficiency. ZSTD is better between these two, that does not automatically make it the better choice for every scenario compared to lighter compression schemes like SNAPPY and LZ4. ZSTD makes sense for data files or exchange data stored on S3. But it's not clear that it would be the best choice for data exchanged over the network. With a high speed and bandwidth network we're better off using less CPU heavy compression. The existing LZ4 exchange compression wasn't turned on by default for pipelined execution because its impact is workload and setup dependent, it makes some queries faster and some slower. This needs benchmarking in an actual Trino cluster rather than JMH or generic compression/decompression speed tests to assess properly.

@wendigo @raunaqmorarka you are right, zstd may not be a silver bullet for all situations~ 
I have modified the pr description, and added a configuration with a session property to switch between different codecs, please cc again, it seems that the CICD error is not related to our pr

File: core/trino-main/src/main/java/io/trino/FeaturesConfig.java [hunk 1]
```diff
@@ -340,6 +343,19 @@ public FeaturesConfig setExchangeCompressionEnabled(boolean exchangeCompressionE
         return this;
     }
 
+    public CompressionKind getCompressionKind()
+    {
+        return compressionKind;
+    }
+
+    @Config("compression-kind")
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/FeaturesConfig.java:351]: "compression-codec"

File: core/trino-main/src/main/java/io/trino/FeaturesConfig.java [hunk 2]
```diff
@@ -340,6 +343,19 @@ public FeaturesConfig setExchangeCompressionEnabled(boolean exchangeCompressionE
...
+    @ConfigDescription("Compression kind in exchanges, spills and page serializations")
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/FeaturesConfig.java:352]: "Compression codec used for data in exchanges and spills"

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java [hunk 3]
```diff
@@ -510,6 +512,12 @@ public SystemSessionProperties(
                         "Enable compression in exchanges",
                         featuresConfig.isExchangeCompressionEnabled(),
                         false),
+                enumProperty(
+                        COMPRESSION_KIND,
+                        "Compression kind in exchanges and page serializations",
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:517]: "Compression codec used for data in exchanges and spills"

File: core/trino-main/src/main/java/io/trino/execution/buffer/CompressionKind.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.buffer;
+
+public enum CompressionKind
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/execution/buffer/CompressionKind.java:16]: `CompressionCodec`

File: core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java [hunk 5]
```diff
@@ -27,20 +38,61 @@ public class PagesSerdeFactory
...
     public PagesSerdeFactory(BlockEncodingSerde blockEncodingSerde, boolean compressionEnabled)
+    {
+        this(blockEncodingSerde, compressionEnabled, LZ4);
+    }
+
+    public PagesSerdeFactory(BlockEncodingSerde blockEncodingSerde, boolean compressionEnabled, CompressionKind compressionKind)
     {
         this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
         this.compressionEnabled = compressionEnabled;
+        this.compressionKind = compressionKind;
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java:52]: `requireNonNull(compressionCodec, "compressionCodec is null")`

File: core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java [hunk 6]
```diff
@@ -27,20 +38,61 @@ public class PagesSerdeFactory
...
+            case SNAPPY:
+                return new SnappyDecompressor();
+            case LZO:
+                return new LzoDecompressor();
+            case LZ4:
+                return new Lz4Decompressor();
+            case ZSTD:
+                return new ZstdDecompressor();
+        }
+        throw new IllegalArgumentException("Unsupported decompression kind" + compressionKind);
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java:96]: missing space between kind and compressionKind

File: core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java [hunk 7]
```diff
@@ -27,20 +38,61 @@ public class PagesSerdeFactory
...
+            case SNAPPY:
+                return new SnappyCompressor();
+            case LZO:
+                return new LzoCompressor();
+            case LZ4:
+                return new Lz4Compressor();
+            case ZSTD:
+                return new ZstdCompressor();
+        }
+        throw new IllegalArgumentException("Unsupported compression kind" + compressionKind);
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java:79]: missing trailing space

File: core/trino-main/src/test/java/io/trino/execution/buffer/TestPagesSerde.java [hunk 8]
```diff
@@ -146,8 +148,8 @@ private void testRoundTrip(List<Type> types, List<Page> pages, int blockSizeInBy
     private void testRoundTrip(List<Type> types, List<Page> pages, boolean compressionEnabled, boolean encryptionEnabled, int blockSizeInBytes)
     {
         Optional<SecretKey> encryptionKey = encryptionEnabled ? Optional.of(createRandomAesEncryptionKey()) : Optional.empty();
-        PageSerializer serializer = new PageSerializer(blockEncodingSerde, compressionEnabled, encryptionKey, blockSizeInBytes);
-        PageDeserializer deserializer = new PageDeserializer(blockEncodingSerde, compressionEnabled, encryptionKey, blockSizeInBytes);
+        PageSerializer serializer = new PageSerializer(blockEncodingSerde, compressionEnabled, new Lz4Compressor(), encryptionKey, blockSizeInBytes);
```
Comment (wendigo) [inline core/trino-main/src/test/java/io/trino/execution/buffer/TestPagesSerde.java:151]: It'd be nice to test all supported compressors with PageSerializer

**Review by wendigo — CHANGES_REQUESTED** [review]

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java (hunk 3 above)
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:517]: Actually I think that we should have 2 separate configuration properties: one for exchanges (data transmission), one for spills (data storage)

File: core/trino-main/src/test/java/io/trino/execution/buffer/TestPagesSerde.java (hunk 8 above)
Comment (hackeryang) [reply core/trino-main/src/test/java/io/trino/execution/buffer/TestPagesSerde.java]: > It'd be nice to test all supported compressors with PageSerializer

@wendigo I have added tests about all compressors, they seemed passed, please cc again~

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java (hunk 3 above)
Comment (hackeryang) [reply core/trino-main/src/main/java/io/trino/SystemSessionProperties.java]: > Actually I think that we should have 2 separate configuration properties: one for exchanges (data transmission), one for spills (data storage)

Yes i haved added a session property and a configuraion for exchanges, and a configuration for spills~ The `FileSingleSpillerFactory` doesn't have a session property api

File: core/trino-main/src/main/java/io/trino/execution/buffer/PageDeserializer.java [hunk 9]
```diff
@@ -62,14 +61,15 @@ public class PageDeserializer
...
+            Decompressor decompressor,
             Optional<SecretKey> encryptionKey,
             int blockSizeInBytes)
     {
         this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
         requireNonNull(encryptionKey, "encryptionKey is null");
         encryptionKey.ifPresent(secretKey -> checkArgument(is256BitSecretKeySpec(secretKey), "encryptionKey is expected to be an instance of SecretKeySpec containing a 256bit key"));
         input = new SerializedPageInput(
-                compressionEnabled ? Optional.of(new Lz4Decompressor()) : Optional.empty(),
+                compressionEnabled ? Optional.ofNullable(decompressor) : Optional.empty(),
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/execution/buffer/PageDeserializer.java:72]: if compression is enabled we should expect non-null decompressor so:

```suggestion
                compressionEnabled ? Optional.of(requireNonNull(decompressor, "decompressor is null")) : Optional.empty(),
```

File: core/trino-main/src/main/java/io/trino/execution/buffer/PageSerializer.java [hunk 10]
```diff
@@ -66,14 +63,15 @@ public class PageSerializer
...
+            Compressor compressor,
             Optional<SecretKey> encryptionKey,
             int blockSizeInBytes)
     {
         this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
         requireNonNull(encryptionKey, "encryptionKey is null");
         encryptionKey.ifPresent(secretKey -> checkArgument(is256BitSecretKeySpec(secretKey), "encryptionKey is expected to be an instance of SecretKeySpec containing a 256bit key"));
         output = new SerializedPageOutput(
-                compressionEnabled ? Optional.of(new Lz4Compressor()) : Optional.empty(),
+                compressionEnabled ? Optional.ofNullable(compressor) : Optional.empty(),
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/execution/buffer/PageSerializer.java:74]:
```suggestion
                compressionEnabled ? Optional.of(requireNonNull(compressor, "compressor is null")) : Optional.empty(),
```

File: core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java [hunk 11]
```diff
@@ -27,20 +38,61 @@ public class PagesSerdeFactory
...
     {
-        return new PageDeserializer(blockEncodingSerde, compressionEnabled, encryptionKey, SERIALIZED_PAGE_DEFAULT_BLOCK_SIZE_IN_BYTES);
+        return new PageDeserializer(blockEncodingSerde, compressionEnabled, createDecompressor(compressionCodec), encryptionKey, SERIALIZED_PAGE_DEFAULT_BLOCK_SIZE_IN_BYTES);
+    }
+
+    public static Compressor createCompressor(CompressionCodec compressionCodec)
+    {
+        switch (compressionCodec) {
+            case NONE:
+                return null;
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java:69]: I don't like the fact that we are returning null here.

What's the sense of the combination `compressionEnabled == true && compressionCodec == NONE` ?

Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java:69]: Let's drop NONE and use the `compressionEnabled` information that states explicitly that compression is not used

**Review by wendigo — CHANGES_REQUESTED** [review]

File: core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java (hunk 11 above)
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java:69]: something along these lines: https://github.com/wendigo/trino/commit/c2a7f64b0e870224921a1f45d24eb6fd4bdee93d#diff-6e2a6eb5c7e2eba83c56db26e3868fcc61a0263a479bc052dd67d33ba864d3afR65

File: core/trino-main/src/main/java/io/trino/execution/buffer/CompressionCodec.java [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.buffer;
+
+public enum CompressionCodec
+{
+    NONE, LZO, SNAPPY, LZ4, ZSTD
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/execution/buffer/CompressionCodec.java:18]: What is the use of `NONE` if we have separate flag to enable/disable compression

File: core/trino-main/src/main/java/io/trino/execution/buffer/PageSerializer.java [hunk 13]
```diff
@@ -94,13 +92,13 @@ private static class SerializedPageOutput
             extends SliceOutput
     {
         private static final int INSTANCE_SIZE = instanceSize(SerializedPageOutput.class);
-        // TODO: implement getRetainedSizeInBytes in Lz4Compressor
-        private static final int COMPRESSOR_RETAINED_SIZE = toIntExact(instanceSize(Lz4Compressor.class) + sizeOfIntArray(Lz4RawCompressor.MAX_TABLE_SIZE));
+        // TODO: implement getRetainedSizeInBytes in Lz4Compressor and ZstdCompressor
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/execution/buffer/PageSerializer.java:95]: this is prerequisite - otherwise we regress on accuracy of memory management.

**Review by losipiuk — COMMENTED** [review]
Nice

File: core/trino-main/src/main/java/io/trino/execution/buffer/CompressionCodec.java (hunk 12 above)
Comment (hackeryang) [reply core/trino-main/src/main/java/io/trino/execution/buffer/CompressionCodec.java]: > What is the use of `NONE` if we have separate flag to enable/disable compression

You are right, NONE is useless in this case, I copied the enum from `io.trino.orc.metadata.CompressionKind` at the beginning~ I have deleted the NONE value

File: core/trino-main/src/main/java/io/trino/execution/buffer/PageSerializer.java (hunk 13 above)
Comment (hackeryang) [reply core/trino-main/src/main/java/io/trino/execution/buffer/PageSerializer.java]: > this is prerequisite - otherwise we regress on accuracy of memory management.

I have added back the computation of Lz4RawCompressor.MAX_TABLE_SIZE, thanks for the advice~

File: core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java (hunk 11 above)
Comment (hackeryang) [reply core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java]: > something along these lines: [wendigo@c2a7f64#diff-6e2a6eb5c7e2eba83c56db26e3868fcc61a0263a479bc052dd67d33ba864d3afR65](https://github.com/wendigo/trino/commit/c2a7f64b0e870224921a1f45d24eb6fd4bdee93d#diff-6e2a6eb5c7e2eba83c56db26e3868fcc61a0263a479bc052dd67d33ba864d3afR65)

Thank you very much for the codes~ yes the NONE enum is useless in this case, I just copied it from `io.trino.orc.metadata.CompressionKind` at the beginning~ I have deleted it this time

File: core/trino-main/src/main/java/io/trino/spiller/NodeSpillConfig.java [hunk 14]
```diff
@@ -14,16 +14,21 @@
...
 
+import static io.trino.execution.buffer.CompressionCodec.LZ4;
+
 public class NodeSpillConfig
 {
     private DataSize maxSpillPerNode = DataSize.of(100, DataSize.Unit.GIGABYTE);
     private DataSize queryMaxSpillPerNode = DataSize.of(100, DataSize.Unit.GIGABYTE);
 
     private boolean spillCompressionEnabled;
+    private CompressionCodec spillCompressionCodec = LZ4;
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/spiller/NodeSpillConfig.java:31]: Compression codec option for spill to disk should be introduced in a separate commit

File: core/trino-main/src/main/java/io/trino/spiller/NodeSpillConfig.java [hunk 15]
```diff
@@ -14,16 +14,21 @@
...
 import jakarta.validation.constraints.NotNull;
 
+import static io.trino.execution.buffer.CompressionCodec.LZ4;
+
 public class NodeSpillConfig
 {
     private DataSize maxSpillPerNode = DataSize.of(100, DataSize.Unit.GIGABYTE);
     private DataSize queryMaxSpillPerNode = DataSize.of(100, DataSize.Unit.GIGABYTE);
 
     private boolean spillCompressionEnabled;
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/spiller/NodeSpillConfig.java:30]: We should deprecate the boolean flag and use the enum, NONE should be the new replacement for boolean flag. This would consistent with how `io.trino.plugin.hive.HiveCompressionOption` works

**Comment by hackeryang** [comment]
@wendigo @losipiuk Thank you for your advice~ I have modified some codes, and the 2 test errors seemed not related to our PR, please cc again~

File: core/trino-main/src/main/java/io/trino/execution/buffer/CompressionCodec.java [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.buffer;
+
+public enum CompressionCodec
+{
+    LZO, SNAPPY, LZ4, ZSTD
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/execution/buffer/CompressionCodec.java:18]: What's the rationale for supporting LZO and SNAPPY here ?
LZ4 offers less aggressive compression at small CPU cost while ZSTD offers aggressive compression at high CPU cost. LZO and SNAPPY are somewhere in between, when would we use either of those two ? We support them in file formats because the standard for those formats requires it. For our own internal usage we should offer fewer choices which fit some use case.

File: core/trino-main/src/main/java/io/trino/FeaturesConfig.java [hunk 17]
```diff
@@ -81,6 +83,7 @@ public class FeaturesConfig
      * default value is overwritten for fault tolerant execution in {@link #applyFaultTolerantExecutionDefaults()}}
      */
     private boolean exchangeCompressionEnabled;
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/FeaturesConfig.java:85]: We should deprecate the boolean flag and instead use NONE enum value to disable compression.

File: core/trino-main/src/main/java/io/trino/FeaturesConfig.java [hunk 18]
```diff
@@ -81,6 +83,7 @@ public class FeaturesConfig
...
+    private CompressionCodec exchangeCompressionCodec = LZ4;
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/FeaturesConfig.java:86]: I ran TPCDS sf1000 benchmark on unpartitioned hive tables with ZSTD exchange compression in pipelined execution mode. The setup is a 6 node cluster of r6g.8xlarge machines on AWS.

[TPCDS unpartitioned parquet zstd exchange.pdf](https://github.com/trinodb/trino/files/13839896/TPCDS.unpartitioned.parquet.zstd.exchange.pdf)

Screenshot 2024-01-05 at 2 18 59 PM

Even though there is a big reduction in internal network data exchanged, there are still regressions in latency due to the high CPU cost of ZSTD compression.
I think this trade-off of more aggressive compression at high CPU cost doesn't make much sense for pipelined execution mode where we exchange data over internal network directly.
This trade off might make sense for filesystem based exchange in fault tolerant execution mode, but that also needs to be assessed with a benchmark.
@losipiuk can we do these changes in way where this config becomes specific to filesystem exchange in FTE ?

File: core/trino-main/src/main/java/io/trino/execution/buffer/CompressionCodec.java (hunk 16 above)
Comment (electrum) [reply core/trino-main/src/main/java/io/trino/execution/buffer/CompressionCodec.java:18]: Agreed. LZ4 should perform better and compress as well as SNAPPY. In other words, LZ4 should be strictly superior to SNAPPY, so there is no need to support it.

File: core/trino-main/src/main/java/io/trino/FeaturesConfig.java (hunk 18 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/FeaturesConfig.java:86]: Benchmark results with FTE
[FTE unpartitioned hive parquet sf1k zstd compression.pdf](https://github.com/trinodb/trino/files/13870656/FTE.unpartitioned.hive.parquet.sf1k.zstd.compression.pdf)

Screenshot 2024-01-09 at 1 26 54 PM

FTE is using LZ4 exchange compression by default already. Switching to ZSTD improves some queries but it's significantly worse overall. So I'm not seeing a compelling reason to add this option even for FTE.

Comment (hackeryang) [reply core/trino-main/src/main/java/io/trino/FeaturesConfig.java]: > Benchmark results with FTE [FTE unpartitioned hive parquet sf1k zstd compression.pdf](https://github.com/trinodb/trino/files/13870656/FTE.unpartitioned.hive.parquet.sf1k.zstd.compression.pdf)
> 
> Screenshot 2024-01-09 at 1 26 54 PM
> FTE is using LZ4 exchange compression by default already. Switching to ZSTD improves some queries but it's significantly worse overall. So I'm not seeing a compelling reason to add this option even for FTE.

You are right, according to the benchmark results, ZSTD should only be used in clusters with little network bandwidth.

We found that some gaming analytics customers of our company have clusters with strong CPU and poor network cards, maybe in this condition we can use it~  

I saw that `Huawei OpenLookeng`(originated from Trino 350) uses ZSTD in exchanges, but it was for their datacenter connector:  
image

**Comment by hackeryang** [comment]
@raunaqmorarka @electrum Thank you very much for the benchmarks and advice~ I have separated two commits for exchanges and spills, deprecated `boolean compressionEnabled` to use enum `CompressionCodec.NONE`, please review again.

I didn't figure out whether I should modify the serialize size in the test:  
image

File: core/trino-main/src/main/java/io/trino/FeaturesConfig.java [hunk 19]
```diff
@@ -328,18 +331,34 @@ public FeaturesConfig setSpillMaxUsedSpaceThreshold(double spillMaxUsedSpaceThre
...
         this.exchangeCompressionEnabled = exchangeCompressionEnabled;
         return this;
     }
 
+    public CompressionCodec getExchangeCompressionCodec()
+    {
+        return exchangeCompressionCodec;
+    }
+
+    @Config("exchange.compression-codec")
```
[the remark this answers is not in the document]
Comment (hackeryang) [reply core/trino-main/src/main/java/io/trino/FeaturesConfig.java:354]: Thanks for the advice, I have added them in `properties-exchange.md` and `properties-spilling.md`

**Comment by hackeryang** [comment]
Relevant slack messages:  https://app.slack.com/client/TFKPX15NC/CP1MUNEUX

File: docs/src/main/sphinx/admin/properties-spilling.md [hunk 20]
```diff
@@ -70,7 +70,15 @@ Limit for memory used for unspilling a single aggregation operator instance.
 - **Type:** {ref}`prop-type-boolean`
 - **Default value:** `false`
 
-Enables data compression for pages spilled to disk.
+Enables data compression for pages spilled to disk. It may be replaced by `spill-compression-codec` in future versions.
```
[the remark this answers is not in the document]
Comment (hackeryang) [reply docs/src/main/sphinx/admin/properties-spilling.md]: @raunaqmorarka Thanks for the suggestions~ I have modified relevant codes and docs, please cc again  
The CI failures seemed not related to our PR, I saw the former merged PRs also have CI failures

File: core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java [hunk 21]
```diff
@@ -26,21 +32,49 @@ public class PagesSerdeFactory
...
+    public static Optional<Compressor> createCompressor(CompressionCodec compressionCodec)
+    {
+        return Optional.ofNullable(switch (compressionCodec) {
+            case NONE:
+                yield null;
+            case LZ4:
+                yield new Lz4Compressor();
+            case ZSTD:
+                yield new ZstdCompressor();
+            default:
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java:62]: default branch is unnecessary here

File: core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java [hunk 22]
```diff
@@ -26,21 +32,49 @@ public class PagesSerdeFactory
...
+    public static Optional<Decompressor> createDecompressor(CompressionCodec compressionCodec)
+    {
+        return Optional.ofNullable(switch (compressionCodec) {
+            case NONE:
+                yield null;
+            case LZ4:
+                yield new Lz4Decompressor();
+            case ZSTD:
+                yield new ZstdDecompressor();
+            default:
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java:76]: default branch is unnecessary here

**Review by raunaqmorarka — APPROVED** [review]

**Comment by hackeryang** [comment]
It seems CI has been passed(except the irrelevant errors), @raunaqmorarka can you please help merge, thanks~

**Comment by raunaqmorarka** [comment]
> It seems CI has been passed(except the irrelevant errors), @raunaqmorarka can you please help merge, thanks~

Please fix the compile error on the first commit

**Comment by hackeryang** [comment]
> > It seems CI has been passed(except the irrelevant errors), @raunaqmorarka can you please help merge, thanks~
> 
> Please fix the compile error on the first commit

@raunaqmorarka I have fixed the error, please cc again~

**Comment by raunaqmorarka** [comment]
> > > It seems CI has been passed(except the irrelevant errors), @raunaqmorarka can you please help merge, thanks~
> > 
> > 
> > Please fix the compile error on the first commit
> 
> @raunaqmorarka I have fixed the error, please cc again~

We still need a fix to the failing test TestJdbcConnection.testSession

**Comment by hackeryang** [comment]
> > > > It seems CI has been passed(except the irrelevant errors), @raunaqmorarka can you please help merge, thanks~
> > > 
> > > 
> > > Please fix the compile error on the first commit
> > 
> > 
> > @raunaqmorarka I have fixed the error, please cc again~
> 
> We still need a fix to the failing test TestJdbcConnection.testSession

@raunaqmorarka Thanks for pointing this out~ I have fixed this test failure, please review again~  

The error this time is about Kudu, seems not related to our PR

File: core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java [hunk 23]
```diff
@@ -26,21 +32,45 @@ public class PagesSerdeFactory
...
+    public static Optional<Compressor> createCompressor(CompressionCodec compressionCodec)
+    {
+        return Optional.ofNullable(switch (compressionCodec) {
+            case NONE:
+                yield null;
+            case LZ4:
+                yield new Lz4Compressor();
+            case ZSTD:
+                yield new ZstdCompressor();
+        });
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java:62]: it'd be more readable to return from switch:


```suggestion
        switch (compressionCodec) {
            case NONE:
                return Optional.empty();
            case LZ4:
                return Optional.of(new Lz4Compressor());
            case ZSTD:
                return Optional.of(new ZstdCompressor());
        };
```

File: core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java [hunk 24]
```diff
@@ -26,21 +32,45 @@ public class PagesSerdeFactory
...
+    public static Optional<Decompressor> createDecompressor(CompressionCodec compressionCodec)
+    {
+        return Optional.ofNullable(switch (compressionCodec) {
+            case NONE:
+                yield null;
+            case LZ4:
+                yield new Lz4Decompressor();
+            case ZSTD:
+                yield new ZstdDecompressor();
+        });
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java:74]:
```suggestion
        switch (compressionCodec) {
            case NONE:
                return Optional.empty();
            case LZ4:
                return Optional.of(new Lz4Decompressor());
            case ZSTD:
                return Optional.of(new ZstdDecompressor());
        });
```

**Review by wendigo — APPROVED** [review]
LGTM % two comments

File: core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java (hunk 24 above)
Comment (hackeryang) [reply core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java]: @wendigo Thanks for the advice~ I have modified relevant codes, please cc again.

Mr raunaq wanted to remove the `case default:` branch, but it will cause a syntax error if we use `switch (compressionCodec)`, so I finally used the `return switch()` syntax and lambda expressions suggested by Intellij IDEA~

Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeFactory.java]: Looks good now :)
